### PR TITLE
Refactor/peptide var pos aware

### PR DESCRIPTION
**Pull request by b-schubert**
- Refactoring of Peptide, Protein, Transcript, Variant
- Peptides can now get positional informations of their Variants per Protein position
 - Getter function calling the Parent Protein Object which handles the request
- Disentangled some relationships between Variant<->Proteins,Transcript
  - Variants do not save the Transcript offsets anymore. This is moved to the transcript and protein objects
- Adjusted generator functions and internal functions accordingly, as well as generator tests
- All internal indices consistently start now at 0

### CODE REVIEW

File: Fred2/Core/Generator.py [hunk 1]
```diff
@@ -23,49 +25,37 @@
...
-    :param transId_old:
-    :param transId_new:
-    """
-    #TODO as this is an manipulator doing an assignment a clearer signature nomenclature would be left- and righthand_side
-    #TODO doc for intended use case
-    for v in vars:
-        offset = v.offsets[transId_old]
-        v.offsets[transId_new] = offset
+#symbol for reverse complement
+REVERS = "-"
```
Comment (mwalzer) [inline Fred2/Core/Generator.py:29]: :+1:

File: Fred2/Core/Generator.py [hunk 2]
```diff
@@ -23,49 +25,37 @@
...
 
 
-#TODO for _incorp... : shouldnt the offset be determined by the variant and _NOT_ be given as argument - otherwise the argument name should be renamed to something like offset_from_previous or transcript_current_offset
-
-def _incorp_snp(seq, var, transId, offset):
+def _incorp_snp(seq, var, transId, pos, offset):
```
Comment (mwalzer) [inline Fred2/Core/Generator.py:32]: makes it def. more flexible to add pos in the header

File: Fred2/Core/Transcript.py [hunk 3]
```diff
@@ -37,29 +37,43 @@ def __init__(self, _seq, _gene_id="unknown", _transcript_id=None, _vars=None):
...
     def __getitem__(self, index):
         """
         Overrides :meth:`Bio.Seq.Seq.__getitem__` (from Biopython)
 
+         Allows only simple slicing (i.e. start < stop)
+
         :param int index: position 
         :returns: (Transcript) -- A Transcript consisting of the single
-        letter at position :attr:`index`.
+        letter at position :attr:`index` or a new sliced Transcript (following Bio.Seqs definition)
```
Comment (mwalzer) [inline Fred2/Core/Transcript.py:57]: should be fine

File: Fred2/test/TestGenerator.py [hunk 4]
```diff
@@ -164,28 +151,26 @@ def test_heterozygous_variants(self):
...
-        #             transToVar.setdefault(trans_id, []).append(v)
+    #     # def test_varinat_reader(self):
+    #     #     vars = read_annovar_exonic("../IO/snp_annot_donor.txt.exonic_variant_function", gene_filter=[])
+    #     #     print vars, len(vars)
+    #     #     trans = list(generate_transcripts_from_variants(vars, self.db_adapter))
+    #     #     print trans
+    #     #     transToVar = {}
+    #     #     for v in vars:
+    #     #         for trans_id in v.coding.iterkeys():
+    #     #             transToVar.setdefault(trans_id, []).append(v)
```
Comment (mwalzer) [inline Fred2/test/TestGenerator.py:176]: and stay commented out! :smile:

File: Fred2/test/TestGenerator.py [hunk 5]
```diff
@@ -20,48 +20,39 @@ def setUp(self):
         self.non_syn_hetero_snp = Variant("COSM1122493", VariationType.SNP, "X",
                                           49111949, "G", "T",
                                           {"NM_001114377": MutationSyntax( \
-                                              "NM_001114377", 757, 218, "", "")
+                                              "NM_001114377", 756, 217, "", "")
```
Comment (mwalzer) [inline Fred2/test/TestGenerator.py:23]: GRCh38?

File: Fred2/Core/Variant.py [hunk 6]
```diff
@@ -104,48 +103,29 @@ def get_transcript_offset(self):
...
-        :param str transcriptId: A transcript_id
-        :return: (int) -- transcript position
-        :raises: KeyError
+        returns the annotated transcript position
+        :param (str) transID: the transcript ID of interest
+        :return: int - the annotated transcript position of the given transcript ID
+        :raise KeyError: if variant is not annotated to the given transcript ID
         """
-        trans_id = trans_variant_id.split(":FRED2_")[0]
+        trID = transID.split(":FRED_")[0]
```
Comment (mwalzer) [inline Fred2/Core/Variant.py:113]: mh, okay no separate member for now - I guess that would entail too much change

File: Fred2/Core/Transcript.py [hunk 7]
```diff
@@ -37,29 +37,43 @@ def __init__(self, _seq, _gene_id="unknown", _transcript_id=None, _vars=None):
...
+            if index.step:
+                slice = set(xrange(start, step, stop))
+            else:
+                slice = set(xrange(start, stop))
+            _vars = {pos-start: v for pos, v in self.vars.iteritems() if pos in slice}
+            trans_id = self.transcript_id+":"+str(Transcript.newid())
+            seq = str(self)[index]
+            t = Transcript(seq, _gene_id=self.gene_id, _transcript_id=trans_id)
+            t.vars = _vars
+            return t
```
Comment (mwalzer) [inline Fred2/Core/Transcript.py:90]: do we have a usecase for the slice yet?

Comment (mwalzer) [inline Fred2/Core/Transcript.py:90]: FS influencing the slice should be (imho) considered.

File: Fred2/Core/Generator.py [hunk 8]
```diff
@@ -165,27 +159,28 @@ def _generate_combinations(tId, vs, seq, usedVs, offset):
...
                 # update the transcript variant id
-                old_trans = generate_peptides_from_variants.transOff
                 generate_peptides_from_variants.transOff += 1
                 transOff = generate_peptides_from_variants.transOff
-                _update_var_offset(usedVs, tId+":FRED2_%i"%old_trans, tId+":FRED2_%i"%(transOff))
-
-                seq, offset = _incorp.get(v.type, lambda a, b, c, d: (a, d))(seq, v, tId+":FRED2_%i"%transOff, offset)
+                pos = vs.coding[tId].tranPos + offset
+                usedVs[pos] = v
+                offset = _incorp.get(v.type, lambda a, b, c, d, e: e)(seq, v, tId+":FRED2_%i"%transOff, pos, offset)
```
Comment (mwalzer) [inline Fred2/Core/Generator.py]: Is this then still necessary? unsure about the use of '...FRED2_' in variant.py@113

File: Fred2/Core/Transcript.py (hunk 7 above)
Comment (b-schubert) [inline Fred2/Core/Transcript.py:90]: Fixed in commit 2a706d4

File: Fred2/Core/Generator.py (hunk 8 above)
Comment (b-schubert) [inline Fred2/Core/Generator.py]: Removed in commit cddf98c

**Pull request merged**